File: agents/seller_agent.py
import json
import os
from web3 import Web3

class PythonSellerAgent:
    def __init__(self, w3, signer_account, contract_address, abi):
        self.w3 = w3
        self.account = signer_account
        self.contract = w3.eth.contract(address=contract_address, abi=abi)

    def ensure_stake(self, min_stake_eth=0.01):
        min_stake_wei = self.w3.to_wei(min_stake_eth, 'ether')
        rep = self.contract.functions.getSellerReputation(self.account.address).call()
        current_stake = rep[0]
        print(f"[Python Seller Agent] Current Stake: {self.w3.from_wei(current_stake, 'ether')} ETH")

        if current_stake < min_stake_wei:
            needed = min_stake_wei - current_stake
            print(f"[Python Seller Agent] Depositing {self.w3.from_wei(needed, 'ether')} ETH stake...")
            tx = self.contract.functions.depositStake().build_transaction({
                'from': self.account.address,
                'value': needed,
                'nonce': self.w3.eth.get_transaction_count(self.account.address),
                'gas': 200000,
                'gasPrice': self.w3.eth.gas_price
            })
            signed = self.account.sign_transaction(tx)
            tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
            self.w3.eth.wait_for_transaction_receipt(tx_hash)
            print(f"[Python Seller Agent] Stake deposit confirmed. Tx: {tx_hash.hex()}")

    def create_listing(self, teaser, plaintext, price_eth, expiry_hours=24, category="Falsifiable", resolution_days=7):
        price_wei = self.w3.to_wei(price_eth, 'ether')
        commitment_hash = self.w3.solidity_keccak(['string'], [plaintext])
        
        # Current block timestamp
        latest_block = self.w3.eth.get_block('latest')
        now = latest_block['timestamp']
        expiry = now + expiry_hours * 3600
        resolution = now + resolution_days * 86400

        print(f"[Python Seller Agent] Posting listing: '{teaser}'")
        tx = self.contract.functions.createListing(
            commitment_hash,
            teaser,
            price_wei,
            expiry,
            0 if category == "Falsifiable" else 1,
            resolution
        ).build_transaction({
            'from': self.account.address,
            'nonce': self.w3.eth.get_transaction_count(self.account.address),
            'gas': 300000,
            'gasPrice': self.w3.eth.gas_price
        })
        signed = self.account.sign_transaction(tx)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
        print(f"[Python Seller Agent] Listing created on-chain! Tx: {tx_hash.hex()}")
        return tx_hash.hex(), commitment_hash

    def reveal_intel(self, purchase_id, plaintext):
        print(f"[Python Seller Agent] Revealing intel for Purchase #{purchase_id}...")
        tx = self.contract.functions.revealIntel(purchase_id, plaintext).build_transaction({
            'from': self.account.address,
            'nonce': self.w3.eth.get_transaction_count(self.account.address),
            'gas': 300000,
            'gasPrice': self.w3.eth.gas_price
        })
        signed = self.account.sign_transaction(tx)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
        print(f"[Python Seller Agent] Proof-of-reveal submitted! Tx: {tx_hash.hex()}")
        return tx_hash.hex()

File: agents/seller_agent.js
const { ethers } = require("ethers");
const fs = require("fs");
const path = require("path");

class SellerAgent {
  constructor(signer, contractAddress, contractAbi) {
    this.signer = signer;
    this.contract = new ethers.Contract(contractAddress, contractAbi, signer);
  }

  async ensureStake(minStakeEth = "0.01") {
    const minStakeWei = ethers.parseEther(minStakeEth);
    const rep = await this.contract.getSellerReputation(this.signer.address);
    console.log(`[Seller Agent] Current stake: ${ethers.formatEther(rep.stake)} ETH`);

    if (rep.stake < minStakeWei) {
      const depositAmount = minStakeWei - rep.stake;
      console.log(`[Seller Agent] Depositing ${ethers.formatEther(depositAmount)} ETH stake to satisfy skin-in-the-game requirement...`);
      const tx = await this.contract.depositStake({ value: depositAmount });
      await tx.wait();
      console.log(`[Seller Agent] Stake deposit confirmed. Tx: ${tx.hash}`);
    }
  }

  async createScoutingListing(intelPackage) {
    const { teaser, plaintextIntel, priceEth, expiryHours, category, resolutionDays } = intelPackage;

    const priceWei = ethers.parseEther(priceEth);
    const commitmentHash = ethers.keccak256(ethers.toUtf8Bytes(plaintextIntel));
    const now = Math.floor(Date.now() / 1000);
    const expiryTimestamp = now + expiryHours * 3600;
    const resolutionTimestamp = now + resolutionDays * 86400;

    console.log(`[Seller Agent] Creating listing: "${teaser}"`);
    console.log(`[Seller Agent] Price: ${priceEth} ETH | Commitment: ${commitmentHash}`);

    const tx = await this.contract.createListing(
      commitmentHash,
      teaser,
      priceWei,
      expiryTimestamp,
      category === "Falsifiable" ? 0 : 1,
      resolutionTimestamp
    );
    const receipt = await tx.wait();

    // Extract listingId from event
    let listingId = 1;
    for (const log of receipt.logs) {
      try {
        const parsed = this.contract.interface.parseLog(log);
        if (parsed && parsed.name === "ListingCreated") {
          listingId = parsed.args.listingId;
          break;
        }
      } catch (e) {}
    }

    console.log(`[Seller Agent] Listing #${listingId} created on-chain! Tx: ${tx.hash}`);
    return { listingId, commitmentHash, plaintextIntel, txHash: tx.hash };
  }

  async revealIntelForPurchase(purchaseId, plaintextIntel) {
    console.log(`[Seller Agent] Revealing intel for purchase #${purchaseId}...`);
    const tx = await this.contract.revealIntel(purchaseId, plaintextIntel);
    const receipt = await tx.wait();
    console.log(`[Seller Agent] Intel revealed on-chain! Tx: ${tx.hash}`);
    return tx.hash;
  }
}

module.exports = SellerAgent;
